### ISSUE TITLE
Implicitly convert f64 to f32 if LHS is constant

### ISSUE THREAD

**Issue by MaxGraey** [AssemblyScript/assemblyscript#1732]
Result context type of `0.5 + x` expression infer as `f64` while
Result context type of `x + 0.5` expression infer as `f32` if `x` has `f32` type.

Example:
```ts
export function test(x: f32): f32 {
  let a = 0.5 + x;
  return Mathf.sqrt(a);
}
```
output:
```ts
ERROR AS200: Conversion from type 'f64' to 'f32' requires an explicit cast.
```
However `a = x + 0.5` or `Mathf.sqrt(0.5 + x)` will compile without any errors and it little bit confused.

I propose implicitly convert RHS from `f64` to `f32` if it's constant.

cc @dcodeIO

**Comment by dcodeIO** [comment]
Yeah, what happens there is that the compiler tries to infer: It first looks at the type annotation of `let a` to obtain the initial contextual type, but since it is omitted, it starts inferring with `auto`, which is essentially `i32`. It then evaluates the expression, here `0.5 + x`, and starts at the LHS. It sees a floating point literal, and now checks whether it can be compiled as `i32` (`auto`), but it can't, so it falls back to `f64`. It then visits the RHS, where the contextual type is now `f64` instead of `auto`, seeing that `x` is of type `f32` and can be implicitly converted to `f64`. The resulting type of `a` is hence `f64`.

I believe there are already a few places in the compiler where it tries to avoid this outcome when literals are in play, and it makes sense to me to attempt to apply it to more binary expressions as well.
